lib/ansible/galaxy/token.py

```diff
@@ -86,9 +98,13 @@ def get(self):
         # TODO: handle auth errors
 
         data = json.loads(to_text(resp.read(), errors='surrogate_or_strict'))
+        token_data = json.loads(
+            base64.b64decode(data['id_token'].split('.')[1] + "==")
+        )
```

Comment:
This incantation looks out of place. It leaks into a low-level abstraction of decoding and parsing the internals.

I'd move at least the last instruction into a pure function so it'd look something like
```python
bearer_token_struct = parse_bearer_token(resp.read())
```
It could return a data structure with both the access token and the id token, if it replaces both instructions.

I don't have visibility into what the HTTP response structure it, but often this data also includes some checksum/signature. Such a function could also verify the integrity and raise an error if it's corrupted somehow.

Maybe it could also normalize the time fields into datetime objects.

Reply:
Good suggestions, thanks. I implemented signature checking. I'm not sure I can assume the JSON web key sets url is relative to the auth URL, so I added a new server config option to achieve that in case the auth_url does not match AH - maybe not backportable now, but it's better I think.

I also fixed normalizing the datetime fields in parse_bearer_token.

Need to figure out how to get sanity tests passing, but I think this is close to done now.